{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp core"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# core\n",
    "\n",
    "> Core functions for analyzing CRISPR base editor screens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "def get_most_severe_mutation_type(mut_types):\n",
    "    \"\"\"\n",
    "    Annotate the most severe mutation type of the\n",
    "    predicted mutation types for a guide sequence\n",
    "    \n",
    "    mut_types: string |\n",
    "    returns: string\n",
    "    \"\"\"\n",
    "    if type(mut_types) == float:\n",
    "        return 'No edits'\n",
    "    elif 'Nonsense' in mut_types:\n",
    "        return 'Nonsense'\n",
    "    elif 'Splice-acceptor' in mut_types or 'Splice-donor' in mut_types:\n",
    "        return 'Splice site'\n",
    "    elif 'Missense' in mut_types:\n",
    "        return 'Missense'\n",
    "    elif 'Intron' in mut_types:\n",
    "        return 'Intron'        \n",
    "    elif 'Silent' in mut_types:\n",
    "        return 'Silent'\n",
    "    elif 'UTR' in mut_types:\n",
    "        return 'UTR'\n",
    "    elif 'Flank' in mut_types:\n",
    "        return 'Flank'\n",
    "    \n",
    "def get_residues(predicted_aa_edits):\n",
    "    \"\"\"\n",
    "    Get the residue positions of all amino acid \n",
    "    predicted edits for a guide sequence\n",
    "    \n",
    "    predicted_aa_edits: string |\n",
    "    returns: string\n",
    "    \"\"\"\n",
    "    res_string = ''\n",
    "    if type(predicted_aa_edits) != float:\n",
    "        edits = [x.strip() for x in re.split(';|,', predicted_aa_edits)]\n",
    "        for edit in edits:\n",
    "            if edit == '(NC)':\n",
    "                res_string += 'non-coding;'\n",
    "            elif edit == '':\n",
    "                continue\n",
    "            else:\n",
    "                res = [x for x in re.findall('\\d+|\\D+', edit) if x.isdigit()][0]\n",
    "                res_string += res\n",
    "                res_string += ';'\n",
    "    return res_string\n",
    "\n",
    "def get_median_residues(mutation_bin, predicted_aa_edits):\n",
    "    \"\"\"\n",
    "    Get the median residue position of predicted \n",
    "    edit positions\n",
    "    \n",
    "    mutation_bin: string |\n",
    "    predicted_aa_edits: string |\n",
    "    returns: float\n",
    "    \"\"\"\n",
    "    residues = get_residues(predicted_aa_edits)\n",
    "    residues = residues.split(';')\n",
    "    if (mutation_bin == 'Missense') or (mutation_bin == 'Nonsense') or (mutation_bin == 'Silent'):\n",
    "        residues = [int(res) for res in residues if res not in ['','non-coding']]\n",
    "        if len(residues) != 0:\n",
    "            return np.median(residues)\n",
    "        else:\n",
    "            return np.nan\n",
    "    else:\n",
    "        return np.nan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To demonstrate the use of these functions, we will first design a base editor tiling library with guides tiling the transcript ENST00000380152 of BRCA2. These guides are annotated with predicted edits using the C>T base editor in the window of nucleotide 4-8."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input</th>\n",
       "      <th>CRISPR Enzyme</th>\n",
       "      <th>Edit Type</th>\n",
       "      <th>Edit Window</th>\n",
       "      <th>Target Assembly</th>\n",
       "      <th>Target Genome Sequence</th>\n",
       "      <th>Target Gene ID</th>\n",
       "      <th>Target Gene Symbol</th>\n",
       "      <th>Target Gene Strand</th>\n",
       "      <th>Target Transcript ID</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM Sequence</th>\n",
       "      <th>sgRNA Target Sequence Start Pos. (global)</th>\n",
       "      <th>sgRNA Orientation</th>\n",
       "      <th>Nucleotide Edits (global)</th>\n",
       "      <th>Guide Edits</th>\n",
       "      <th>Nucleotide Edits</th>\n",
       "      <th>Amino Acid Edits</th>\n",
       "      <th>Mutation Category</th>\n",
       "      <th>Constraint Violations</th>\n",
       "      <th>Note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>SpyoCas9</td>\n",
       "      <td>C-T</td>\n",
       "      <td>4..8</td>\n",
       "      <td>GRCh38 (9606)</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000380152.8</td>\n",
       "      <td>...</td>\n",
       "      <td>TGG</td>\n",
       "      <td>32316449</td>\n",
       "      <td>sense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>SpyoCas9</td>\n",
       "      <td>C-T</td>\n",
       "      <td>4..8</td>\n",
       "      <td>GRCh38 (9606)</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000380152.8</td>\n",
       "      <td>...</td>\n",
       "      <td>AGG</td>\n",
       "      <td>32316462</td>\n",
       "      <td>sense</td>\n",
       "      <td>32316465C&gt;T</td>\n",
       "      <td>C_4</td>\n",
       "      <td>5C&gt;T</td>\n",
       "      <td>Pro2Leu</td>\n",
       "      <td>Missense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>SpyoCas9</td>\n",
       "      <td>C-T</td>\n",
       "      <td>4..8</td>\n",
       "      <td>GRCh38 (9606)</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000380152.8</td>\n",
       "      <td>...</td>\n",
       "      <td>AGG</td>\n",
       "      <td>32316467</td>\n",
       "      <td>antisense</td>\n",
       "      <td>32316479G&gt;A;32316481G&gt;A, 32316483G&gt;A</td>\n",
       "      <td>C_8_6, C_4</td>\n",
       "      <td>19G&gt;A;21G&gt;A, 23G&gt;A</td>\n",
       "      <td>Glu7Lys, Arg8Lys</td>\n",
       "      <td>Missense, Missense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>SpyoCas9</td>\n",
       "      <td>C-T</td>\n",
       "      <td>4..8</td>\n",
       "      <td>GRCh38 (9606)</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000380152.8</td>\n",
       "      <td>...</td>\n",
       "      <td>TGG</td>\n",
       "      <td>32316477</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000380152</td>\n",
       "      <td>SpyoCas9</td>\n",
       "      <td>C-T</td>\n",
       "      <td>4..8</td>\n",
       "      <td>GRCh38 (9606)</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>ENSG00000139618</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000380152.8</td>\n",
       "      <td>...</td>\n",
       "      <td>TGG</td>\n",
       "      <td>32316488</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Input CRISPR Enzyme Edit Type Edit Window Target Assembly  \\\n",
       "0  ENST00000380152      SpyoCas9       C-T        4..8   GRCh38 (9606)   \n",
       "1  ENST00000380152      SpyoCas9       C-T        4..8   GRCh38 (9606)   \n",
       "2  ENST00000380152      SpyoCas9       C-T        4..8   GRCh38 (9606)   \n",
       "3  ENST00000380152      SpyoCas9       C-T        4..8   GRCh38 (9606)   \n",
       "4  ENST00000380152      SpyoCas9       C-T        4..8   GRCh38 (9606)   \n",
       "\n",
       "  Target Genome Sequence   Target Gene ID Target Gene Symbol  \\\n",
       "0           NC_000013.11  ENSG00000139618              BRCA2   \n",
       "1           NC_000013.11  ENSG00000139618              BRCA2   \n",
       "2           NC_000013.11  ENSG00000139618              BRCA2   \n",
       "3           NC_000013.11  ENSG00000139618              BRCA2   \n",
       "4           NC_000013.11  ENSG00000139618              BRCA2   \n",
       "\n",
       "  Target Gene Strand Target Transcript ID  ... PAM Sequence  \\\n",
       "0                  +    ENST00000380152.8  ...          TGG   \n",
       "1                  +    ENST00000380152.8  ...          AGG   \n",
       "2                  +    ENST00000380152.8  ...          AGG   \n",
       "3                  +    ENST00000380152.8  ...          TGG   \n",
       "4                  +    ENST00000380152.8  ...          TGG   \n",
       "\n",
       "  sgRNA Target Sequence Start Pos. (global) sgRNA Orientation  \\\n",
       "0                                  32316449             sense   \n",
       "1                                  32316462             sense   \n",
       "2                                  32316467         antisense   \n",
       "3                                  32316477         antisense   \n",
       "4                                  32316488         antisense   \n",
       "\n",
       "              Nucleotide Edits (global)  Guide Edits    Nucleotide Edits  \\\n",
       "0                                   NaN          NaN                 NaN   \n",
       "1                           32316465C>T          C_4                5C>T   \n",
       "2  32316479G>A;32316481G>A, 32316483G>A   C_8_6, C_4  19G>A;21G>A, 23G>A   \n",
       "3                                   NaN          NaN                 NaN   \n",
       "4                                   NaN          NaN                 NaN   \n",
       "\n",
       "   Amino Acid Edits   Mutation Category Constraint Violations Note  \n",
       "0               NaN                 NaN                   NaN  NaN  \n",
       "1           Pro2Leu            Missense                   NaN  NaN  \n",
       "2  Glu7Lys, Arg8Lys  Missense, Missense                   NaN  NaN  \n",
       "3               NaN                 NaN                   NaN  NaN  \n",
       "4               NaN                 NaN                   NaN  NaN  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "design_df = pd.read_csv('sample_input/crisprbe-guides.txt', sep='\\t')\n",
    "design_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Assign severe mutation bin\n",
    "\n",
    "As noted in the \"Mutation Category\" column, each guide is predicted to make more one or more types of mutations if Cs are present in the editing window. We can then annotate each guide with the most severe mutation bin in the order Nonsense > Splice site > Missense > Intron > Silent > UTR > no edit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA Target Sequence</th>\n",
       "      <th>Mutation Category</th>\n",
       "      <th>Mutation Bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGTAGGTAAAAATGCCTAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGCCTATTGGATCCAAAGAG</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GGCCTCTCTTTGGATCCAAT</td>\n",
       "      <td>Missense, Missense</td>\n",
       "      <td>Missense</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAATGTTGGCCTCTCTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTAAAAATTTCAAAAAATGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sgRNA Target Sequence   Mutation Category Mutation Bin\n",
       "0  TCGTAGGTAAAAATGCCTAT                 NaN     No edits\n",
       "1  TGCCTATTGGATCCAAAGAG            Missense     Missense\n",
       "2  GGCCTCTCTTTGGATCCAAT  Missense, Missense     Missense\n",
       "3  AAAAAATGTTGGCCTCTCTT                 NaN     No edits\n",
       "4  TTAAAAATTTCAAAAAATGT                 NaN     No edits"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "design_df['Mutation Bin'] = design_df['Mutation Category'].apply(get_most_severe_mutation_type)\n",
    "design_df[['sgRNA Target Sequence','Mutation Category','Mutation Bin']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate median residue\n",
    "We can then get the median residue of the predicted edits."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA Target Sequence</th>\n",
       "      <th>Amino Acid Edits</th>\n",
       "      <th>Mutation Category</th>\n",
       "      <th>Mutation Bin</th>\n",
       "      <th>Median Residue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGTAGGTAAAAATGCCTAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGCCTATTGGATCCAAAGAG</td>\n",
       "      <td>Pro2Leu</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GGCCTCTCTTTGGATCCAAT</td>\n",
       "      <td>Glu7Lys, Arg8Lys</td>\n",
       "      <td>Missense, Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>7.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAATGTTGGCCTCTCTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TTAAAAATTTCAAAAAATGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AAGACACGCTGCAACAAAGC</td>\n",
       "      <td>Thr17Ile, Arg18Cys</td>\n",
       "      <td>Missense, Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>17.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TTTTTTTTTTAAATAGATTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TAGGACCAATAAGTCTTAAT</td>\n",
       "      <td>Pro26Leu</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>26.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCAAACCAATTAAGACTTAT</td>\n",
       "      <td>Trp31Ter</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GCAGGTTCAGAATTATAGGG</td>\n",
       "      <td>Glu45Lys</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TCTGCAGGTTCAGAATTATA</td>\n",
       "      <td>Ala47Thr</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TTCTGCAGGTTCAGAATTAT</td>\n",
       "      <td>Ala47Thr</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>47.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TTATGTTCAGATTCTTCTGC</td>\n",
       "      <td>Glu51Lys</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>51.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TGTGGAGTTTTAAATAGGTT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACCTATTTAAAACTCCACAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No edits</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sgRNA Target Sequence    Amino Acid Edits   Mutation Category Mutation Bin  \\\n",
       "0   TCGTAGGTAAAAATGCCTAT                 NaN                 NaN     No edits   \n",
       "1   TGCCTATTGGATCCAAAGAG             Pro2Leu            Missense     Missense   \n",
       "2   GGCCTCTCTTTGGATCCAAT    Glu7Lys, Arg8Lys  Missense, Missense     Missense   \n",
       "3   AAAAAATGTTGGCCTCTCTT                 NaN                 NaN     No edits   \n",
       "4   TTAAAAATTTCAAAAAATGT                 NaN                 NaN     No edits   \n",
       "5   AAGACACGCTGCAACAAAGC  Thr17Ile, Arg18Cys  Missense, Missense     Missense   \n",
       "6   TTTTTTTTTTAAATAGATTT                 NaN                 NaN     No edits   \n",
       "7   TAGGACCAATAAGTCTTAAT            Pro26Leu            Missense     Missense   \n",
       "8   TCAAACCAATTAAGACTTAT            Trp31Ter            Nonsense     Nonsense   \n",
       "9   GCAGGTTCAGAATTATAGGG            Glu45Lys            Missense     Missense   \n",
       "10  TCTGCAGGTTCAGAATTATA            Ala47Thr            Missense     Missense   \n",
       "11  TTCTGCAGGTTCAGAATTAT            Ala47Thr            Missense     Missense   \n",
       "12  TTATGTTCAGATTCTTCTGC            Glu51Lys            Missense     Missense   \n",
       "13  TGTGGAGTTTTAAATAGGTT                 NaN                 NaN     No edits   \n",
       "14  ACCTATTTAAAACTCCACAA                 NaN                 NaN     No edits   \n",
       "\n",
       "    Median Residue  \n",
       "0              NaN  \n",
       "1              2.0  \n",
       "2              7.5  \n",
       "3              NaN  \n",
       "4              NaN  \n",
       "5             17.5  \n",
       "6              NaN  \n",
       "7             26.0  \n",
       "8             31.0  \n",
       "9             45.0  \n",
       "10            47.0  \n",
       "11            47.0  \n",
       "12            51.0  \n",
       "13             NaN  \n",
       "14             NaN  "
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "design_df['Median Residue'] = design_df.apply(lambda x: get_median_residues(x['Mutation Bin'], x['Amino Acid Edits']), axis=1)\n",
    "design_df[['sgRNA Target Sequence','Amino Acid Edits','Mutation Category','Mutation Bin','Median Residue']].head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
