---

title: core


keywords: fastai
sidebar: home_sidebar

summary: "Core functions for analyzing CRISPR base editor screens"
description: "Core functions for analyzing CRISPR base editor screens"
nb_path: "00_core.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 00_core.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_most_severe_mutation_type" class="doc_header"><code>get_most_severe_mutation_type</code><a href="https://github.com/gpprnd/poola_be/tree/master/poola_be/core.py#L9" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_most_severe_mutation_type</code>(<strong><code>mut_types</code></strong>)</p>
</blockquote>
<p>Annotate the most severe mutation type of the
predicted mutation types for a guide sequence</p>
<p>mut_types: string |
returns: string</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_residues" class="doc_header"><code>get_residues</code><a href="https://github.com/gpprnd/poola_be/tree/master/poola_be/core.py#L34" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_residues</code>(<strong><code>predicted_aa_edits</code></strong>)</p>
</blockquote>
<p>Get the residue positions of all amino acid
predicted edits for a guide sequence</p>
<p>predicted_aa_edits: string |
returns: string</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_median_residues" class="doc_header"><code>get_median_residues</code><a href="https://github.com/gpprnd/poola_be/tree/master/poola_be/core.py#L56" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_median_residues</code>(<strong><code>mutation_bin</code></strong>, <strong><code>predicted_aa_edits</code></strong>)</p>
</blockquote>
<p>Get the median residue position of predicted
edit positions</p>
<p>mutation_bin: string |
predicted_aa_edits: string |
returns: float</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To demonstrate the use of these functions, we will first design a base editor tiling library with guides tiling the transcript ENST00000380152 of BRCA2. These guides are annotated with predicted edits using the C&gt;T base editor in the window of nucleotide 4-8.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">design_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;sample_input/crisprbe-guides.txt&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">design_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Input</th>
      <th>CRISPR Enzyme</th>
      <th>Edit Type</th>
      <th>Edit Window</th>
      <th>Target Assembly</th>
      <th>Target Genome Sequence</th>
      <th>Target Gene ID</th>
      <th>Target Gene Symbol</th>
      <th>Target Gene Strand</th>
      <th>Target Transcript ID</th>
      <th>...</th>
      <th>PAM Sequence</th>
      <th>sgRNA Target Sequence Start Pos. (global)</th>
      <th>sgRNA Orientation</th>
      <th>Nucleotide Edits (global)</th>
      <th>Guide Edits</th>
      <th>Nucleotide Edits</th>
      <th>Amino Acid Edits</th>
      <th>Mutation Category</th>
      <th>Constraint Violations</th>
      <th>Note</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ENST00000380152</td>
      <td>SpyoCas9</td>
      <td>C-T</td>
      <td>4..8</td>
      <td>GRCh38 (9606)</td>
      <td>NC_000013.11</td>
      <td>ENSG00000139618</td>
      <td>BRCA2</td>
      <td>+</td>
      <td>ENST00000380152.8</td>
      <td>...</td>
      <td>TGG</td>
      <td>32316449</td>
      <td>sense</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ENST00000380152</td>
      <td>SpyoCas9</td>
      <td>C-T</td>
      <td>4..8</td>
      <td>GRCh38 (9606)</td>
      <td>NC_000013.11</td>
      <td>ENSG00000139618</td>
      <td>BRCA2</td>
      <td>+</td>
      <td>ENST00000380152.8</td>
      <td>...</td>
      <td>AGG</td>
      <td>32316462</td>
      <td>sense</td>
      <td>32316465C&gt;T</td>
      <td>C_4</td>
      <td>5C&gt;T</td>
      <td>Pro2Leu</td>
      <td>Missense</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ENST00000380152</td>
      <td>SpyoCas9</td>
      <td>C-T</td>
      <td>4..8</td>
      <td>GRCh38 (9606)</td>
      <td>NC_000013.11</td>
      <td>ENSG00000139618</td>
      <td>BRCA2</td>
      <td>+</td>
      <td>ENST00000380152.8</td>
      <td>...</td>
      <td>AGG</td>
      <td>32316467</td>
      <td>antisense</td>
      <td>32316479G&gt;A;32316481G&gt;A, 32316483G&gt;A</td>
      <td>C_8_6, C_4</td>
      <td>19G&gt;A;21G&gt;A, 23G&gt;A</td>
      <td>Glu7Lys, Arg8Lys</td>
      <td>Missense, Missense</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ENST00000380152</td>
      <td>SpyoCas9</td>
      <td>C-T</td>
      <td>4..8</td>
      <td>GRCh38 (9606)</td>
      <td>NC_000013.11</td>
      <td>ENSG00000139618</td>
      <td>BRCA2</td>
      <td>+</td>
      <td>ENST00000380152.8</td>
      <td>...</td>
      <td>TGG</td>
      <td>32316477</td>
      <td>antisense</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ENST00000380152</td>
      <td>SpyoCas9</td>
      <td>C-T</td>
      <td>4..8</td>
      <td>GRCh38 (9606)</td>
      <td>NC_000013.11</td>
      <td>ENSG00000139618</td>
      <td>BRCA2</td>
      <td>+</td>
      <td>ENST00000380152.8</td>
      <td>...</td>
      <td>TGG</td>
      <td>32316488</td>
      <td>antisense</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>5 rows Ã— 23 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Assign-severe-mutation-bin">Assign severe mutation bin<a class="anchor-link" href="#Assign-severe-mutation-bin"> </a></h1><p>As noted in the "Mutation Category" column, each guide is predicted to make more one or more types of mutations if Cs are present in the editing window. We can then annotate each guide with the most severe mutation bin in the order Nonsense &gt; Splice site &gt; Missense &gt; Intron &gt; Silent &gt; UTR &gt; no edit.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">design_df</span><span class="p">[</span><span class="s1">&#39;Mutation Bin&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">design_df</span><span class="p">[</span><span class="s1">&#39;Mutation Category&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">get_most_severe_mutation_type</span><span class="p">)</span>
<span class="n">design_df</span><span class="p">[[</span><span class="s1">&#39;sgRNA Target Sequence&#39;</span><span class="p">,</span><span class="s1">&#39;Mutation Category&#39;</span><span class="p">,</span><span class="s1">&#39;Mutation Bin&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sgRNA Target Sequence</th>
      <th>Mutation Category</th>
      <th>Mutation Bin</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TCGTAGGTAAAAATGCCTAT</td>
      <td>NaN</td>
      <td>No edits</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TGCCTATTGGATCCAAAGAG</td>
      <td>Missense</td>
      <td>Missense</td>
    </tr>
    <tr>
      <th>2</th>
      <td>GGCCTCTCTTTGGATCCAAT</td>
      <td>Missense, Missense</td>
      <td>Missense</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AAAAAATGTTGGCCTCTCTT</td>
      <td>NaN</td>
      <td>No edits</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TTAAAAATTTCAAAAAATGT</td>
      <td>NaN</td>
      <td>No edits</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Calculate-median-residue">Calculate median residue<a class="anchor-link" href="#Calculate-median-residue"> </a></h1><p>We can then get the median residue of the predicted edits.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">design_df</span><span class="p">[</span><span class="s1">&#39;Median Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">design_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">get_median_residues</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;Mutation Bin&#39;</span><span class="p">],</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Amino Acid Edits&#39;</span><span class="p">]),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">design_df</span><span class="p">[[</span><span class="s1">&#39;sgRNA Target Sequence&#39;</span><span class="p">,</span><span class="s1">&#39;Amino Acid Edits&#39;</span><span class="p">,</span><span class="s1">&#39;Mutation Category&#39;</span><span class="p">,</span><span class="s1">&#39;Mutation Bin&#39;</span><span class="p">,</span><span class="s1">&#39;Median Residue&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">15</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sgRNA Target Sequence</th>
      <th>Amino Acid Edits</th>
      <th>Mutation Category</th>
      <th>Mutation Bin</th>
      <th>Median Residue</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TCGTAGGTAAAAATGCCTAT</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>No edits</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TGCCTATTGGATCCAAAGAG</td>
      <td>Pro2Leu</td>
      <td>Missense</td>
      <td>Missense</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>GGCCTCTCTTTGGATCCAAT</td>
      <td>Glu7Lys, Arg8Lys</td>
      <td>Missense, Missense</td>
      <td>Missense</td>
      <td>7.5</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AAAAAATGTTGGCCTCTCTT</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>No edits</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TTAAAAATTTCAAAAAATGT</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>No edits</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>5</th>
      <td>AAGACACGCTGCAACAAAGC</td>
      <td>Thr17Ile, Arg18Cys</td>
      <td>Missense, Missense</td>
      <td>Missense</td>
      <td>17.5</td>
    </tr>
    <tr>
      <th>6</th>
      <td>TTTTTTTTTTAAATAGATTT</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>No edits</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>TAGGACCAATAAGTCTTAAT</td>
      <td>Pro26Leu</td>
      <td>Missense</td>
      <td>Missense</td>
      <td>26.0</td>
    </tr>
    <tr>
      <th>8</th>
      <td>TCAAACCAATTAAGACTTAT</td>
      <td>Trp31Ter</td>
      <td>Nonsense</td>
      <td>Nonsense</td>
      <td>31.0</td>
    </tr>
    <tr>
      <th>9</th>
      <td>GCAGGTTCAGAATTATAGGG</td>
      <td>Glu45Lys</td>
      <td>Missense</td>
      <td>Missense</td>
      <td>45.0</td>
    </tr>
    <tr>
      <th>10</th>
      <td>TCTGCAGGTTCAGAATTATA</td>
      <td>Ala47Thr</td>
      <td>Missense</td>
      <td>Missense</td>
      <td>47.0</td>
    </tr>
    <tr>
      <th>11</th>
      <td>TTCTGCAGGTTCAGAATTAT</td>
      <td>Ala47Thr</td>
      <td>Missense</td>
      <td>Missense</td>
      <td>47.0</td>
    </tr>
    <tr>
      <th>12</th>
      <td>TTATGTTCAGATTCTTCTGC</td>
      <td>Glu51Lys</td>
      <td>Missense</td>
      <td>Missense</td>
      <td>51.0</td>
    </tr>
    <tr>
      <th>13</th>
      <td>TGTGGAGTTTTAAATAGGTT</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>No edits</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>14</th>
      <td>ACCTATTTAAAACTCCACAA</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>No edits</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

